crypto/ocsp/ocsp_client.c

```diff
@@ -3,22 +3,113 @@
 
 #include "internal.h"
 
-OCSP_ONEREQ *OCSP_request_add0_id(OCSP_REQUEST *req, OCSP_CERTID *cid)
-{
-    OCSP_ONEREQ *one = OCSP_ONEREQ_new();
-    if (one == NULL) {
-      return NULL;
+OCSP_ONEREQ *OCSP_request_add0_id(OCSP_REQUEST *req, OCSP_CERTID *cid) {
+  OCSP_ONEREQ *one = OCSP_ONEREQ_new();
+  if (one == NULL) {
+    return NULL;
+  }
+  // Reassign |OCSP_CERTID| allocated by OCSP_ONEREQ_new().
+  OCSP_CERTID_free(one->reqCert);
+  one->reqCert = cid;
+  if (req != NULL && !sk_OCSP_ONEREQ_push(req->tbsRequest->requestList, one)) {
+    one->reqCert = NULL;  // do not free on error
+    OCSP_ONEREQ_free(one);
+    return NULL;
+  }
+  return one;
+}
+
+int OCSP_request_set1_name(OCSP_REQUEST *req, X509_NAME *nm) {
+  GENERAL_NAME *gen = GENERAL_NAME_new();
+  if (gen == NULL) {
+    return 0;
+  }
+  if (!X509_NAME_set(&gen->d.directoryName, nm)) {
+    GENERAL_NAME_free(gen);
+    return 0;
+  }
+  gen->type = GEN_DIRNAME;
+  GENERAL_NAME_free(req->tbsRequest->requestorName);
+  req->tbsRequest->requestorName = gen;
+  return 1;
+}
+
+int OCSP_request_add1_cert(OCSP_REQUEST *req, X509 *cert) {
+  if (req->optionalSignature == NULL) {
+    req->optionalSignature = OCSP_SIGNATURE_new();
+  }
+  OCSP_SIGNATURE *sig = req->optionalSignature;
+  if (sig == NULL) {
+    return 0;
+  }
+  if (cert == NULL) {
+    return 1;
+  }
+
+  if (sig->certs == NULL && (sig->certs = sk_X509_new_null()) == NULL) {
+    return 0;
+  }
+
+  if (!sk_X509_push(sig->certs, cert)) {
+    return 0;
+  }
+  // sk_X509_push takes ownership.
+  X509_up_ref(cert);
+  return 1;
+}
+
+int OCSP_request_sign(OCSP_REQUEST *req,
+                      X509 *signer,
+                      EVP_PKEY *key,
+                      const EVP_MD *dgst,
+                      STACK_OF(X509) *certs, unsigned long flags) {
+  if (req->optionalSignature != NULL) {
+    // OpenSSL lets an |OCSP_REQUEST| be signed twice, regardless of whether
```

Comment:
1. Just curious how did you find this edge case?
2. Do we have a test that attempts to sign twice?
3. Is this worthy of a new error message?

Reply:
1. I was trying to sign the signed OCSP requests test cases we have (`ocsp_request_signed`, `ocsp_request_signed_sha256`, etc.) with `OCSP_request_sign`, but I couldn't get past the ASAN build. Initially I couldn't make sense of it, but I dove a bit deeper and found that the original `req->optionalSignature` was being left behind.
2. If we try signing the signed OCSP requests test cases again, we'll be hitting this and fail. I've included those cases in `kRequestTestVectors`. 
3. I had been pondering that too, I'll add it.